### ISSUE TITLE
How can I access nested model fields for Preload chaining?

### ISSUE THREAD

**Issue by sudopower** [go-gorm/gen#373]
## How can I access nested model fields for Preload chaining?
 The example mentioned on GORM.io works fine but not when using GORM GEN
Nested Model fields are not accessible

## The document you expected this should be explained
https://github.com/go-gorm/gen#nested-preloading
This works fine in GORM but not GORM-GEN

https://gorm.io/docs/preload.html#Nested-Preloading
This works fine in GORM but not GORM-GEN

## Expected answer
A way to preload nested association Fields

Let me explain my situation with renamed examples

```
type Parent struct {
	ID            int32   `gorm:"column:id;type:int;primaryKey;autoIncrement:false" json:"id"`
	Name 	      string `gorm:"column:name;type:varchar(64)" json:"name"`
	Children      []Child `gorm:"foreignKey:ParentID" json:"children"`
}

type Child struct {
	ID            	  int32                `gorm:"column:id;type:int;primaryKey;autoIncrement:true" json:"id"`
	ParentID          *int32               `gorm:"column:parent_id;type:int;" json:"parent_id"`
	GrandSons   	  []GrandSon           `gorm:"foreignKey:ChildID" json:"grand_ss"`
	GrandDaughters    []GrandDaughter      `gorm:"foreignKey:ChildID" json:"grand_ds"`
}

type GrandSon struct {
	ID  	  int32 `gorm:"column:id;type:int;primaryKey;autoIncrement:false;" json:"id"`
	Name 	  string `gorm:"column:name;type:varchar(64)" json:"name"`
	ParentID  int32  `gorm:"column:ChildID;type:int;primaryKey;autoIncrement:false" json:"child_id"`
        Hobbies	  []Hobby `gorm:"foreignKey:ParentID" json:"hobbies"`
}

type GrandDaughter struct {
	ID        int32 `gorm:"column:id;type:int;primaryKey;autoIncrement:false;" json:"id"`
	Name 	  string `gorm:"column:name;type:varchar(64)" json:"name"`
	ParentID  int32  `gorm:"column:ChildID;type:int;primaryKey;autoIncrement:false" json:"child_id"`
        Hobbies	  []Hobby `gorm:"foreignKey:ParentID" json:"hobbies"`
}

type Hobby struct {
	ID        int32 `gorm:"column:id;type:int;primaryKey;autoIncrement:false;" json:"id"`
	Name 	  string `gorm:"column:name;type:varchar(64)" json:"name"`
}


conn := db.GetDB()
var tree 
conn.Where("id= ?",id).
WithContext(c).
Preload("Children.GrandSons").
Preload("Children.GrandDaughters").
First(&tree)
```

This will give me the below result which is as expected
```
{
    "id": 1,
    "name": "john",
    "grand_ss": [
        {
            "id": 1,
            "name": "John",
            "hobbies": [
                {
                    "id": 1,
                    "name": "running"
                }
            ]
        }
    ],
    "grand_ds": [
        {
            "id": 1,
            "name": "Jane",
            "hobbies": [
                {
                    "id": 1,
                    "name": "running"
                }
            ]
        }
    ]
}
```

I assume GORM GEN equivalent would be 
- OPTION 1
	```
	id:= 1
	t := query.Use(db.GetDB()).Parent
	tree, err := t.Where(t.ID.Eq(id)).
	Preload(field.Association). 
	First()
	```
	This only expands the first level of Associations
	```
    {
        "id": 1,
        "name": "john",
        "grand_ss": [
            {
                "id": 1,
                "name": "John",
                "hobbies": null
            }
        ],
        "grand_ds": [
            {
                "id": 1,
                "name": "Jane",
                "hobbies": null
            }
        ]
    }
	```

- OPTION 2
	```
	id:= 1
	t := query.Use(db.GetDB()).Parent
	tree, err := t.Where(t.ID.Eq(id)).
	Preload(t.Children.RelationField). 
	First()
	```

	Again this will only expand the first level of Associations

I see in the GORM a solution is to chain preload with nested structure [here](https://gorm.io/docs/preload.html#Nested-Preloading) and [here](https://github.com/go-gorm/gen#nested-preloading)

But this works only for GORM, I cannot access the nested items in GORM-GEN

##TL;DR
#### It would be great to do something like this. Is there some alternative way of achieving the topmost example response?
```
#example
id:= 1
t := query.Use(db.GetDB()).Parent
tree, err := t.Where(t.ID.Eq(id)).
Preload(t.Children.GrandSons.RelationField). 
Preload(t.Children.GrandDaughters.RelationField). 
First()
```

**Comment by sudopower** [comment]
I'm using GORM instead of GORM-GEN as a workaround, any updates ?

Thanks

**Comment by riverchu** [comment]
in progress

**Comment by riverchu** [comment]
Does those structs you used as example generated by `GEN`? Actually `GEN` already support [chain preload](https://github.com/go-gorm/gen#nested-preloading) association. May you show me the generated query code? For example
```go
func newCustomer(db *gorm.DB) customer {
	_customer := customer{}

	_customer.customerDo.UseDB(db)
	_customer.customerDo.UseModel(&model.Customer{})

	tableName := _customer.customerDo.TableName()
	_customer.ALL = field.NewField(tableName, "*")
	...
	_customer.CreditCards = customerHasManyCreditCards{
		db: db.Session(&gorm.Session{}),

		RelationField: field.NewRelation("CreditCards", "model.CreditCard"),
		Bank: struct {
			field.RelationField
		}{
			RelationField: field.NewRelation("CreditCards.Bank", "model.Bank"),
		},
	}
        ...
}
```

**Comment by sudopower** [comment]
@riverchu  
Thanks for your reply

Yes, it is generated by `GEN` 
My function looks like this.

```
func newFoo(db *gorm.DB) foo {
	_foo := foo{}

	_foo.fooDo.UseDB(db)
	_foo.fooDo.UseModel(&model.Foo{})

	tableName := _foo.fooDo.TableName()
	_foo.ALL = field.NewField(tableName, "*")
	_foo.ID = field.NewInt32(tableName, "id")
	_foo.DefaultBazID = field.NewInt32(tableName, "default_language_id")
	_foo.CountryID = field.NewInt32(tableName, "country_id")
	_foo.CreatedAt = field.NewTime(tableName, "created_at")
	_foo.UpdatedAt = field.NewTime(tableName, "updated_at")
	_foo.Bars = fooHasManyBars{
		db: db.Session(&gorm.Session{}),

		RelationField: field.NewRelation("Bars", "model.Bar"),
	}

	_foo.Bazs = fooHasManyBazs{
		db: db.Session(&gorm.Session{}),

		RelationField: field.NewRelation("Bazs", "model.FooBazAssoc"),
	}

	_foo.Bay = fooHasOneBay{
		db: db.Session(&gorm.Session{}),

		RelationField: field.NewRelation("Bay", "model.Bay"),
	}

	_foo.fillFieldMap()

	return _foo
}
```


Now within "Bars" if I have other associations that I want to expand, I don't see a solution for that.
With GORM I've managed to do it like below -

```
conn := db.GetDB()
var dist *model.Foo = nil
conn.Where("id= ?", id).WithContext(c).Preload("Bar.Beer").Preload("Bazs").Preload("Bay").First(&dist)
```

Which is ok for now, but, is a workaround.

**Comment by riverchu** [comment]
ok, i will give you a example

```go
// models

// Customer mapped from table <customers>
type Customer struct {
	ID          int64          `gorm:"column:id;type:bigint(20) unsigned;primaryKey;autoIncrement:true" json:"id"`
	CreditCards []CreditCard   `gorm:"foreignKey:CustomerRefer" json:"credit_cards"`
}

// CreditCard mapped from table <credit_cards>
type CreditCard struct {
	ID            int64          `gorm:"column:id;type:bigint(20) unsigned;primaryKey;autoIncrement:true" json:"id"`
	CustomerRefer int64          `gorm:"column:customer_refer;type:bigint(20) unsigned;index:fk_customers_credit_cards,priority:1" json:"customer_refer"`
	BankID        int64          `gorm:"column:bank_id;type:bigint(20) unsigned;index:fk_credit_cards_bank,priority:1" json:"bank_id"`
	Bank          Bank           `gorm:"" json:"bank"`
}

// Bank mapped from table <banks>
type Bank struct {
	ID      int64  `gorm:"column:id;type:bigint(20) unsigned;primaryKey;autoIncrement:true" json:"id"`
	Name    string `gorm:"column:name;type:longtext" json:"name"`
	Address string `gorm:"column:address;type:longtext" json:"address"`
	Scale   int64  `gorm:"column:scale;type:bigint(20)" json:"scale"`
}
```

```go
func newCustomer(db *gorm.DB) customer {
	_customer := customer{}

	_customer.customerDo.UseDB(db)
	_customer.customerDo.UseModel(&model.Customer{})

	tableName := _customer.customerDo.TableName()
	_customer.ALL = field.NewField(tableName, "*")
	_customer.ID = field.NewInt64(tableName, "id")
	_customer.CreatedAt = field.NewTime(tableName, "created_at")
	_customer.UpdatedAt = field.NewTime(tableName, "updated_at")
	_customer.DeletedAt = field.NewField(tableName, "deleted_at")
	_customer.BankID = field.NewInt64(tableName, "bank_id")
	_customer.CreditCards = customerHasManyCreditCards{
		db: db.Session(&gorm.Session{}),

		RelationField: field.NewRelation("CreditCards", "model.CreditCard"),
		Bank: struct {
			field.RelationField
		}{
			RelationField: field.NewRelation("CreditCards.Bank", "model.Bank"),
		},
	}

	_customer.fillFieldMap()

	return _customer
}

type customer struct {
	customerDo customerDo

	ALL         field.Field
	ID          field.Int64
	CreatedAt   field.Time
	UpdatedAt   field.Time
	DeletedAt   field.Field
	BankID      field.Int64
	CreditCards customerHasManyCreditCards

	fieldMap map[string]field.Expr
}
```


```go
customer, err = c.WithContext(ctx).Where(c.ID.Eq(1)).Preload(c.CreditCards.Bank).Take()
```
And i got this:
```json
{
    "id":1,
    "credit_cards":[
        {
            "id":2,
            "number":"456",
            "customer_refer":1,
            "bank_id":2,
            "bank":{
                "id":2,
                "name":"Bank2",
                "address":"abc",
                "scale":0
            }
        }
    ]
}
```

**Comment by riverchu** [comment]
maybe you should try this in your first example
```go
 t := query.Use(db.GetDB()).Parent
 tree, err := t.Where(t.ID.Eq(1)).
 Preload(t.Children.Hobbies). 
 First()
```

**Comment by sudopower** [comment]
I read all the tables and their associations from a config, loop over it, and generate all the models and CRUD helpers.
I'll have to make some changes to my flow to be able to implement this suggestion. 
Then it will work. 

Thanks for the prompt replies and all the help :)
Closing the issue!

**Comment by sudopower** [comment]
> maybe you should try this in your first example
> 
> ```go
>  t := query.Use(db.GetDB()).Parent
>  tree, err := t.Where(t.ID.Eq(1)).
>  Preload(t.Children.Hobbies). 
>  First()
> ```

This works thanks!
